File: aoc2023/clojure-solutions/src/clojure_solutions/day01.clj
(ns clojure-solutions.day01
  (:require [clojure.string :as str])
  (:use [clojure-aoc-util.util] :reload))

(defn- solve1 [input]
  (->> input
       (map (partial re-seq #"[1-9]"))
       (map (fn [xs] (parse-long (str/join [(first xs) (last xs)]))))
       sum))

(def replacements
  ["one" "two" "three" "four" "five" "six" "seven" "eight" "nine" ])

(defn- solve2 [input]
  (letfn [(to-digit [[d s]]
            (if (not (empty? d))
              (parse-long d)
              (+ 1 (.indexOf replacements s))))]
    (->> input
         (map (partial re-seq
                       (re-pattern (str/join ["[1-9]|(?=("
                                              (str/join "|" replacements)
                                              "))"]))))
         (map (fn [xs] (+ (* (to-digit (first xs))
                             10)
                          (to-digit (last xs)))))
         sum)))

(defn day01 [part]
  (let [input (str/split-lines (slurp "../inputs/day01.txt"))]
    (case part
      :one (solve1 input)
      :two (solve2 input))))

File: aoc2023/python-solutions/src/day01.py
import re
from functools import reduce
from util import INPUTS


def solve1(inp: str) -> int:
    matches = re.findall(r"[1-9]", inp)
    return int(matches[0]) * 10 + int(matches[-1])


replacements = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]


def solve2(inp: str) -> int:
    def convert(a: str, b: str) -> int:
        if a != "":
            return int(a)
        else:
            return 1 + replacements.index(b)

    regexp = re.compile(
        "([1-9])|(?=(" + reduce((lambda acc, el: acc + "|" + el), replacements) + "))"
    )
    matches = re.findall(regexp, inp)
    return convert(*matches[0]) * 10 + convert(*matches[-1])


def parse() -> list[str]:
    with open(f"{INPUTS}/day01.txt", "r") as inp:
        return [x.strip() for x in inp.readlines()]


def day01() -> tuple[int, int]:
    inp = parse()
    return sum([solve1(x) for x in inp]), sum([solve2(x) for x in inp])


assert day01() == (54331, 54518)

if __name__ == "__main__":
    print(day01())
